File: src/leetCode/java/Solution103.java
package leetCode.java;

import java.util.*;

//BFS O(V) O(V)
class Solution103 {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if(root == null) return new ArrayList<>();

        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        List<List<Integer>> res = new ArrayList<>();
        boolean goRight = true;
        while(!q.isEmpty()){
            int size = q.size();
            List<Integer> level = new ArrayList<>(size);
            for(int i=0; i<size; i++)
                level.add(0);

            for(int i=0; i<size; i++) {
                TreeNode node = q.poll();
                int index = goRight ? i : (size - 1 - i);
                level.set(index, node.val);
                if(node.left != null) q.offer(node.left);
                if(node.right != null) q.offer(node.right);
            }
            res.add(level);
            goRight = !goRight;
        }
        return res;
    }
}

File: c++/Solution103.cpp
//BFS O(V) O(V)
class Solution103 {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(!root) return {};

        queue<TreeNode*> q;
        q.push(root);
        bool goRight = true;//方向->
        vector<vector<int>> res;
        while(!q.empty()){
            int size = q.size();
            vector<int> level(size);
            for(int i=0; i<size; ++i){
                TreeNode* node = q.front(); q.pop();
                int index = goRight ? i : (size - 1 - i);
                level[index] = node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
            res.push_back(level);
            goRight = !goRight;
        }
        return res;
    }
};